### ISSUE TITLE
chunkwm gets in a loop waiting for an XPC Service to finish launching

### ISSUE THREAD

**Issue by kylef** [koekeishiya/chunkwm#489]
While debugging chunkwm I spot sometimes chunkwm gets in a loop waiting for an XPC Service to "finish launching".

```
2018-08-15 17:54:07 DEBUG: 77642:com.apple.WebKit.Networking waiting for application to be ready for notifications
2018-08-15 17:54:07 DEBUG: 77642:com.apple.WebKit.Networking waiting for application to be ready for notifications
2018-08-15 17:54:08 DEBUG: 77642:com.apple.WebKit.Networking waiting for application to be ready for notifications
2018-08-15 17:54:08 DEBUG: 77642:com.apple.WebKit.Networking waiting for application to be ready for notifications
2018-08-15 17:54:08 DEBUG: 77642:com.apple.WebKit.Networking waiting for application to be ready for notifications
2018-08-15 17:54:08 DEBUG: 77642:com.apple.WebKit.Networking waiting for application to be ready for notifications
2018-08-15 17:54:08 DEBUG: 77642:com.apple.WebKit.Networking waiting for application to be ready for notifications
```

This appears to continue until either the XPC Service exits or perhaps chunkwm restarts, I am not entirely sure on the case. But I can see these messages in the log for a long time (20 mins +) on repeat causing chunkwm to waste some CPU cycles in this infinite loop.

In this case, the XPC service is  `/System/Library/Frameworks/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.Networking.xpc/Contents/MacOS/com.apple.WebKit.Networking` which I assume is an XPC service for Safari (but I could be wrong).

Looking at the code, there is a check if the application is "interactive" before this is invoked inside `CarbonApplicationEventHandler` it seems that the XPC Service is incorrect assumed as interactive. While this may be a bug elsewhere than chunkwm, I think perhaps we should work around it in chunkwm because of the infinite dispatch loop it causes if that is indeed the case.

The `GetProcessInformation` function seems to be responsible for populating the state about the process and if it is interactive. I cannot find any documentation for this or where it is defined. Is this a reverse engineered function from Carbon? 

chunkwm: c4f349e
macOS: 10.13.6 (17G65)

**Comment by koekeishiya** [comment]
It's documented in the carbon process manager: http://mirror.informatimago.com/next/developer.apple.com/documentation/Carbon/Reference/Process_Manager/prmref_main/FunctionGroupIndex.html

The problem is that some applications that should b be interactive, are reported as if they are not interactive, and so I tried to cheat the system by quering yet another API to see if the API was finished launcing, but some applications seem to never report finished state, and so it loops forever..

**Comment by kylef** [comment]
> API to see if the API was finished launcing, but some applications seem to never report finished state

In this case that won't work for XPC services are not NSApplication-based.

> Some applications do not post this notification (applications that do not rely on NSApplication) and so are never reported as finished launching. [isFinishedLaunching](https://developer.apple.com/documentation/appkit/nsrunningapplication/1532002-isfinishedlaunching)

Do you have any idea how I could get into a state where I have an "interactive" XPC Service? I've tried to build an XPC Service with type "Application" and "User" but they are not being caught by chunkwm at all.

**Comment by koekeishiya** [comment]
Not really no. I don't know of any means to know if the application rely on the NSApplication  services or if it is an XPC service.

The problem described here happens with applications that use an activation-policy of:
> NSApplicationActivationPolicyAccessory
>> The application does not appear in the Dock and does not have a menu bar, but it may be activated programmatically or by clicking on one of its windows.

Some applications use this mode, but DO NOT respond to the "isFinishedLaunching" event.
I've seen this happen with e.g webkit when opening a Safari tab or in Qutebrowser.

We could revert to only accepting
> NSApplicationActivationPolicyRegular
>> The application is an ordinary app that appears in the Dock and may have a user interface.

but then applications such as macOS *System Information* and likely some others are not going to be watched correctly.

Another problem is that applications can call a function at runtime to convert between a GUI based application and a console based application `TransformProcessType(&psn, kProcessTransformToForegroundApplication);` but afaik there is no way for us to be notified about when this process transformation occur. This seems to be what is happening to *mpv* when ran from the terminal  (#108).

**Comment by koekeishiya** [comment]
Fixed on master. Did not find any API to query whether it is an XPC service or not. Simply added a timeout after 15seconds. If the application does not respond confirming that it was ready launching within this timeframe, we simply give up and ignore it.

I also checked using a private function *CGSGetConnectionIDForPSN(...)* and that always returned failure until [application finishedLaunching] was true, so there appears to be a correlation regarding these.

The function I mentioned above returns the process'es connection to the WindowServer. It makes sense that there is no connection created until after the process has finished launching.
Unfortunately there does not seem to be a way to get to know in advance if the application is going to have a valid connection to the WindowServer or not, so this does not actually do anything for us except for gaining some knowledge into how this all works under the hood, as we would still have to poll, and in the end we don't know if the function is ever going to return a valid connection to us.

I chose 15seconds as the timeout duration, as I've noticed some applications to take quite a bit of time to start properly, and these were previously ignored by chunkwm (and this is on a relatively performant mbp). I feel like 15seconds should be more than enough for a proper application to report that it has finished launching, and if it does not, it will simply be ignored by chunkwm.

As long as we have to use an exhaustive / polling strategy, I don't see a better way to handle this situation, so I'll be closing this issue.
